enterprise/views.py

```diff
@@ -605,36 +646,72 @@ def post(self, request, enterprise_uuid, course_id):
-
-        if matching_modes[0].get('premium', False):
-            # Premium modes should be directed to the ecommerce flow
+        # Verify that the request user belongs to the enterprise against the
+        # provided `enterprise_uuid` and if user does not belong to the
+        # enterprise then redirect the user to the ecommerce flow in LMS.
+        enterprise_customer_user = get_enterprise_customer_user(request.user.id, enterprise_uuid)
+        if not enterprise_customer_user:
             return redirect(LMS_START_PREMIUM_COURSE_FLOW_URL.format(course_id=course_id))
 
-        with transaction.atomic():
-            # Create the Enterprise backend database records for this course
-            # enrollment, then create the real enrollment in the LMS. Finally,
-            # redirect the user to the courseware page, where they'll be prompted
-            # for data sharing consent if they need to be.
-            ec_user, __ = EnterpriseCustomerUser.objects.get_or_create(
-                user_id=request.user.id,
-                enterprise_customer=enterprise_customer,
+        selected_course_mode_name = request.POST.get('course_mode')
+        selected_course_mode = None
+        for course_mode in course_modes:
+            if course_mode['mode'] == selected_course_mode_name:
+                selected_course_mode = selected_course_mode_name
+                break
+
+        if not selected_course_mode:
+            return self.get_enterprise_course_enrollment_page(request, enterprise_customer, course, course_modes)
+
+        # For the audit course modes `audit`, `honor`, where DSC is not
+        # required manually enroll the learner and redirect the learner to the
+        # courseware page.
+        if not selected_course_mode.get('premium') and not user_requires_data_sharing_consent(enterprise_customer_user):
+            with transaction.atomic():
+                # Create the Enterprise backend database records for this course
+                # enrollment.
+                EnterpriseCourseEnrollment.objects.get_or_create(
+                    enterprise_customer_user=enterprise_customer_user,
+                    course_id=course_id,
+                )
+
+                # In case of Audit course modes enroll the learner directly through
+                # enrollment API client and redirect the learner to dashboard.
+                client = EnrollmentApiClient()
+                client.enroll_user_in_course(request.user.username, course_id, selected_course_mode_name)
+
+            return redirect(LMS_COURSEWARE_URL.format(course_id=course_id))
+
+        # For the audit `audit`, `honor` or premium course modes `Verified`,
+        # `Prof Ed` where DSC is required, redirect the learner to course
+        # specific DSC with enterprise UUID from where the learner will be
+        # directed to the ecommerce flow after providing DSC.
+        if user_requires_data_sharing_consent(enterprise_customer_user):
```

Comment:
This should be an `elif`; not an entirely new block. Additionally, can we do it without an extra DB call?